### PR TITLE
Add Tutorials for Building a Language from A-Z

### PR DESCRIPTION
**Pull request by montymxb**
This PR adds 7 new tutorials to our documentation. These tutorials are focused on guiding a user from getting a default language produced via the yeoman generator, all the way to a complete language running in the web with generation. These tutorials are all written around MiniLogo to provide a concrete example from one guide to the next.

This is quite a lot of tutorials, and a lot of new material, so reviews, corrections, improvements would be greatly appreciated 📚 .

In particular, I'm really interested in knowing how this *flows*, and whether the progression of tutorials (and their topics) makes sense.

EDIT: Revised from 'guides' to 'tutorials', given the strict continuity between each of these.

### CODE REVIEW

**Comment by Lotes** [comment]
I was wondering why only one guide is linked to the documentation (inside of the preview deployment).
Forgotten or by intent?

File: hugo/content/guides/building_an_extension.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+In this guide we'll be going over how to build a VSIX extension (VSCode extension) for your Langium-based language. This will allow providing LSP support in VSCode for your language. We'll assume that you've already looked at the previous guides, particularly the bundling guide, so that you're ready to build an extension. At this point we assume that your language is also working, and there are no issues running `npm run langium:generate` or `npm run build`. If there are, you'll want to correct those first.
+
+## Setting up the Scripts
+
+To get started, you'll want to have a language expressed in Langium, such as [Lox](https://github.com/langium/langium-lox) or [MiniLogo](https://github.com/langium/langium-minilogo). If you have been following along with these guides, you should already have something ready. If you don't you can also use the default language generated by the yeoman generator for Langium, presented in the [getting started](/docs/getting-started/) section.
+
+Regardless of what you're working with, you'll want to make sure you have the following scripts in your **package.json**.
+
+```json
+{
```
Comment (Lotes) [inline hugo/content/guides/building_an_extension.md:15]: I think the scripts section is nice to have.
The Linting script, is it important to add?
Here we could focus and just explain the `esbuild` command.

File: hugo/content/guides/building_an_extension.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+You may also notice that your extension may not have an icon to start with. This is a small thing that we can fix quickly. This is as simple as adding a small PNG icon somewhere in your project repo, such as the root. You'll also want to set the **icon** property in your package.json with the relative path to this icon. When you regenerate your extension & reinstall it, you should get an icon that is the same as the one that you packaged it with.
```
Comment (Lotes) [inline hugo/content/guides/building_an_extension.md:41]: Add an icon? That is nice and may deserves an own headline?
These points can be also part of an agenda at the beginning of the entire text

**Comment by montymxb** [comment]
Just a quick note, the netlify preview seems to be missing these guides. This seems to be related to #65.

**Comment by msujew** [comment]
@montymxb They seem to be [visible](https://deploy-preview-86--langium-website.netlify.app/guides/building_an_extension/), the ToC entries are just missing.

**Review by msujew — COMMENTED** [review]
I'm wondering whether we should have this kind of content under a "Tutorial" section. It's quite interconnected content and requires for the most part to reuse known principles from other tutorials. That way it also wouldn't polute the ToC of the guides as much.

File: hugo/content/guides/building_an_extension.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+```bash
+vsce package
+```
+
+You should now see a VSIX extension file in the root of your project. The name of this file will correspond with the **name** and **version** properties listed in your **package.json**. For MiniLogo, this produced **minilogo-0.1.0.vsix**.
+
+For installing the extension, you can right click the extension file, and select "Install VSIX Extension" at the bottom of the list. You should see a small indication that your VSIX extension has been successfully installed. You can verify this by going to your extensions tab and looking at the enabled extensions, where you should find the name of your language (again corresponding to the **name** property in your package.json). Assuming the extension is enabled and working correctly, you can open any file that ends in the extensions registered for your language, and you should immediately observe the syntax highlighting kicking in. Interaction with your language should show that syntax errors are recognized, and other LSP functionalities are working as intended (such as renaming of symbols).
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/guides/building_an_extension.md:39]: Good suggestion, I can put that in sometime next week

File: hugo/content/guides/building_an_extension.md (hunk 2 above)
Comment (montymxb) [reply hugo/content/guides/building_an_extension.md:41]: That can be added. I wonder if geekdoc supports an auto-ToC feature, could help with organization at the start.

**Comment by montymxb** [comment]
@msujew I have no problem with that, would save rewriting parts to make them more general. They do make sense as they are connected together, so that would make sense.

File: hugo/content/guides/building_an_extension.md (hunk 1 above)
Comment (montymxb) [reply hugo/content/guides/building_an_extension.md:15]: Dropped the linting script.

File: hugo/content/guides/building_an_extension.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/guides/building_an_extension.md:39]: Broke up these sections with plenty of helpful images 🖼️

File: hugo/content/guides/building_an_extension.md (hunk 2 above)
Comment (montymxb) [reply hugo/content/guides/building_an_extension.md:41]: This has it's own sub-header now. But the local TOC I have to look again. Some hugo themes will automatically inject the table of contents at the start, others need a flag to do so, and some don't have it

Comment (montymxb) [reply hugo/content/guides/building_an_extension.md]: Local TOC is also setup now 👍

**Comment by montymxb** [comment]
Side note: Removed the draft flag from the remaining pages, that was why they weren't showing up on the netlify preview (production build).

File: hugo/content/guides/code-bundling.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ---
 title: "Code Bundling"
-weight: 0
+weight: 4
```
Comment (msujew) [inline hugo/content/guides/code-bundling.md:3]: Please revert this change before merging.

**Review by msujew — APPROVED** [review]
Looks great 👍

**Review by kaisalmen — CHANGES_REQUESTED** [review]
Superb work! You guide the reader nicely through all the topics. 
I found some minor things (typos) and some things that need adjustment due to recent changes.

File: hugo/content/tutorials/writing_a_grammar.md [hunk 5]
```diff
@@ -0,0 +1,306 @@
+---
+title: "Writing a Grammar"
+weight: 0
+---
+
+{{< toc format=html >}}
+
+In this tutorial we will be talking about writing a grammar for your langauge in Langium. As a motivating example, we'll be describing how to write a grammar for the MiniLogo language. If you're not familiar with MiniLogo, it's a smaller implementation of the Logo programming language. Logo itself is a lot like Turtle from Python. Ultimately, we'll be using MiniLogo to express drawing instructions that can be used to draw on a canvas.
+
+We've already written an implemention of [MiniLogo on Github using Langium](https://github.com/langium/langium-minilogo). This tutorial will be following along with this project, bs walking through the grammar implementation step by step. Later tutorials will also follow along with MiniLogo to create an easy to follow series.
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/writing_a_grammar.md:10]: Oh 😆 , nope that should be `by walking`

File: hugo/content/tutorials/writing_a_grammar.md [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+As an aside, our version of MiniLogo will be an *approximation* of Dr. Walkingshaw's version. We won't adhere to it completely, and we won't be incorporating some elements, such as variable declarations.
+
+To get started sketching the grammar we'll be using the **Hello World** example from the yeoman generator. You can read about how to get this setup in the [Getting Started](/docs/getting-started/) section of our docs. We'll be working with a fresh from the generator using only the defaults, and building up from that. We'll begin by modifying the default grammar file, and updating it to work for MiniLogo. You can find this file under **src/language-server/hello-world.langium** in your new project. If you used a name other than the default, the file will still be there, but using your custom name instead.
+
+We'll be overriding the existing langium grammar file completely, so delete the old contents before we begin.
+
+The first line that we'll then add is the declaration of our grammar.
+
+```antlr
+grammar MiniLogo
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/writing_a_grammar.md:46]: There's an issue with the syntax highlighting that Hugo uses. There's no direct way to correct it, but we *can* change the syntax that we highlight for. Might be something that we can find that works a bit better (or not highlighting to disable this kind of stuff).

File: hugo/content/tutorials/writing_a_grammar.md [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+hidden terminal SL_COMMENT:     /\/\/[^\n\r]*/;
+```
+
+And that's it, we're all set writing up the grammar for MiniLogo. To verify that we correctly implemented the grammar with no problems, we can run the following command in the project root:
+
+```bash
+npm run langium:generate
+```
+
+The generation should finish successfully, indicating that our grammar doesn't have any errors in it. In some cases, you may get warnings, but these shouldn't prevent the generation from completing successfully. Also, when we're referring to the generation, we're talking about the construction of the following from your grammar:
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/writing_a_grammar.md:301]: You may get warnings related to unused / unreachable rules (such as for dummy rules), as an example. But they're soft, and don't cause issues with the build. I'll clarify that in a revision later today.

File: hugo/content/tutorials/writing_a_grammar.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/writing_a_grammar.md:46]: Leaving this as is, since https://github.com/langium/langium-website/pull/111 should resolve this.

File: hugo/content/tutorials/writing_a_grammar.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/writing_a_grammar.md:301]: Added a small note about this in an upcoming commit.

File: hugo/content/tutorials/langium_and_monaco.md [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+import { MonacoEditorLanguageClientWrapper } from './monaco-editor-wrapper/index.js';
+import { buildWorkerDefinition } from "./monaco-editor-workers/index.js";
+
+buildWorkerDefinition('./monaco-editor-workers/workers', new URL('', window.location.href).href, false);
+
+MonacoEditorLanguageClientWrapper.addMonacoStyles('monaco-editor-styles');
+MonacoEditorLanguageClientWrapper.addCodiconTtf();
+```
+
+Then, we'll want to instantiate our language client wrapper, and give it some unique identifier (42 in this case). We'll also need to get a handle for the editor configuration, and set the current language id to `minilogo`. This should match the id of the language that will be recognized by our language server.
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/langium_and_monaco.md:206]: Dropped that part of the sentence altogether.

File: hugo/content/tutorials/langium_and_monaco.md [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+
+// keep a reference to a promise for when the editor is finished starting, we'll use this to setup the canvas on load
+const startingPromise = client.startEditor(document.getElementById("monaco-editor-root"));
+
+window.addEventListener("resize", () => client.updateLayout());
+```
+
+Note the `startingPromise` that's returned from `startEditor`. We're not using this yet, but it will be important for our setup in the next tutorial.
+
+## Serving via Express
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/langium_and_monaco.md:305]: Can be done in a future change, but it's a good idea.

File: hugo/content/tutorials/generation_in_the_web.md [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+let posY = 0;
+```
+
+And let's begin dispatching each of our commands. To do this, we'll setup an interval that repeatedly shifts the top element from our list of commands, dispatches, it and repeats. Once we're out of commands to dispatch, we'll clear the interval. Feel free to adjust the delay (or remove it entirely) in your version.
+
+```js
+// use the command list to execute each commmand with a small delay
+const id = setInterval(() => {
+    if (cmds.length > 0) {
+        dispatchCommand(cmds.shift(), context);
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/generation_in_the_web.md:421]: Removed notions of dispatch entirely, settled on evaluation, more accurate here.

File: hugo/content/tutorials/_index.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+These tutorials revolve around the MiniLogo language, implemented in Langium. They describe how to implement MiniLogo from the ground up, and how to customize the various features of MiniLogo.
+
+By working through each of these tutorials, you'll be learning about:
+
+- writing a grammar in Langium
+- implementing validation for your language
+- customizing a CLI for your tooling
+- writing up a simple generator
+- building a VSCode extension
+- running Langium in the web with Monaco
```
[the remark this answers is not in the document]
Comment (montymxb) [reply hugo/content/tutorials/_index.md:17]: Went with changing the title to `Langium + Monaco Editor` as well, displays nicely on the side bar still.

File: hugo/content/tutorials/langium_and_monaco.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kaisalmen) [reply hugo/content/tutorials/langium_and_monaco.md:206]: Sounds good.

**Review by kaisalmen — APPROVED** [review]
LGTM 👍
